core/src/avm2/globals/json.rs

```diff
@@ -0,0 +1,443 @@
+            let mut arr: Vec<Option<Value<'gc>>> = Vec::with_capacity(js_arr.len());
+            for (key, val) in js_arr.iter().enumerate() {
+                let val = deserialize_json_inner(activation, val.clone(), reviver)?;
+                let mapped_val = reviver.map_or(Ok(val.clone()), |reviver| {
+                    reviver.call(None, &[key.into(), val], activation)
+                })?;
+                arr.push(Some(mapped_val));
+            }
+            let storage = ArrayStorage::from_storage(arr);
+            let array = ArrayObject::from_storage(activation, storage)?;
+            array.into()
+        }
+    })
+}
+
+fn deserialize_json<'gc>(
+    activation: &mut Activation<'_, 'gc, '_>,
+    json: JsonValue,
+    reviver: Option<Object<'gc>>,
+) -> Result<Value<'gc>, Error> {
+    let val = deserialize_json_inner(activation, json, reviver)?;
+    reviver.map_or(Ok(val.clone()), |reviver| {
+        reviver.call(None, &["".into(), val], activation)
+    })
+}
+
+/// This is a custom generator backend for the json crate that allows modifying the indentation character.
+/// This generator is used in `JSON.stringify` when a string is passed to the `space` parameter.
+struct FlashStrGenerator<'a> {
+    code: Vec<u8>,
+    indent_str: &'a str,
+    indent_size: u16,
+}
+
+impl<'a> FlashStrGenerator<'a> {
+    fn new(indent_str: &'a str) -> Self {
+        Self {
+            code: Vec::with_capacity(1024),
+            indent_str,
+            indent_size: 0,
+        }
+    }
+
+    pub fn consume(self) -> String {
+        // SAFETY: JSON crate should never generate invalid UTF-8
+        unsafe { String::from_utf8_unchecked(self.code) }
+    }
+}
+
+impl<'a> JsonGenerator for FlashStrGenerator<'a> {
+    type T = Vec<u8>;
+
+    #[inline(always)]
+    fn write(&mut self, slice: &[u8]) -> std::io::Result<()> {
+        self.code.extend_from_slice(slice);
+        Ok(())
+    }
+
+    #[inline(always)]
+    fn write_char(&mut self, ch: u8) -> std::io::Result<()> {
+        self.code.push(ch);
+        Ok(())
+    }
+
+    #[inline(always)]
+    fn get_writer(&mut self) -> &mut Vec<u8> {
+        &mut self.code
+    }
+
+    #[inline(always)]
+    fn write_min(&mut self, slice: &[u8], _: u8) -> std::io::Result<()> {
+        self.code.extend_from_slice(slice);
+        Ok(())
+    }
+
+    fn new_line(&mut self) -> std::io::Result<()> {
+        self.code.push(b'\n');
+        for _ in 0..self.indent_size {
+            self.code.extend_from_slice(self.indent_str.as_bytes());
+        }
+        Ok(())
+    }
+
+    fn indent(&mut self) {
+        self.indent_size += 1;
+    }
+
+    fn dedent(&mut self) {
+        self.indent_size -= 1;
+    }
+}
+
+enum Replacer<'gc> {
+    Function(FunctionObject<'gc>),
+    PropList(ArrayObject<'gc>),
+}
+
+struct AvmSerializer<'gc> {
+    /// This object stack will be used to detect circular references and return an error instead of a panic.
+    obj_stack: Vec<*const ObjectPtr>,
```

Comment:
I've seen this was a simple `Object<'gc>` before, so I'm curious - why the change to pointer?

Reply:
The reason was because I assumed that storing the pointer would be cheaper than storing the whole object (which contains extra metadata). Although, now that I think about it, the call to `as_ptr()` may be more expensive than I thought because it needs to use the objects vtable, so I will probably revert this change

Reply:
I just tested it on desktop, and actually `obj1 == obj2` also compiles terribly due to virtual call :( The approach with `*const ObjectPtr` avoids _some_ repeated virtual calls, so maybe let's stay with that for now? But it's not that perf-critical 
 either and might be improved in the future too... up to you :)